core/src/snapshot_packager_service.rs

```diff
@@ -109,6 +138,58 @@ impl SnapshotPackagerService {
     pub fn join(self) -> thread::Result<()> {
         self.t_snapshot_packager.join()
     }
+
+    /// Get the next snapshot package to handle
+    ///
+    /// Look through the snapshot package channel to find the highest priority one to handle next.
+    /// If there are no snapshot packages in the channel, return None.  Otherwise return the
+    /// highest priority one.  Unhandled snapshot packages with slots GREATER-THAN the handled one
+    /// will be re-enqueued.  The remaining will be dropped.
+    ///
+    /// Also return the number of snapshot packages initially in the channel, and the number of
+    /// ones re-enqueued.
+    fn get_next_snapshot_package(
+        snapshot_package_sender: &Sender<SnapshotPackage>,
+        snapshot_package_receiver: &Receiver<SnapshotPackage>,
+    ) -> Option<(
+        SnapshotPackage,
+        /*num outstanding snapshot packages*/ usize,
+        /*num re-enqueued snapshot packages*/ usize,
+    )> {
+        let mut snapshot_packages: Vec<_> = snapshot_package_receiver.try_iter().collect();
+        // `select_nth()` panics if the slice is empty, so return if that's the case
+        if snapshot_packages.is_empty() {
+            return None;
+        }
+        let snapshot_packages_len = snapshot_packages.len();
+        debug!("outstanding snapshot packages ({snapshot_packages_len}): {snapshot_packages:?}");
+
+        snapshot_packages.select_nth_unstable_by(
+            snapshot_packages_len - 1,
+            snapshot_package::cmp_snapshot_packages_by_priority,
+        );
+        // SAFETY: We know `snapshot_packages` is not empty, so its len is >= 1,
+        // therefore there is always an element to pop.
+        let snapshot_package = snapshot_packages.pop().unwrap();
+        let handled_snapshot_package_slot = snapshot_package.slot();
+        // re-enqueue any remaining snapshot packages for slots GREATER-THAN the snapshot package
+        // that will be handled
+        let num_re_enqueued_snapshot_packages = snapshot_packages
+            .into_iter()
+            .filter(|snapshot_package| snapshot_package.slot() > handled_snapshot_package_slot)
```

Comment:
What does slot() return for incremental package type? Is it the baseline or the ending slot? Do we want to filter out any incrementals that don't attempt to build off the latest full snapshot?